spatialdata/_core/_spatialdata_ops.py

```diff
@@ -291,154 +289,90 @@ def _describe_paths(paths: list[list[Union[int, str]]]) -> str:
-            if isinstance(region, list):
-                all_region_keys.append(region_key)
-            else:
-                assert (
-                    len(all_region_keys) == 0
-                    or len(set(all_region_keys)) == 1
-                    and all_region_keys[0] == MERGED_TABLES_REGION_KEY
-                )
-                table.obs[MERGED_TABLES_REGION_KEY] = region
-                all_region_keys.append(MERGED_TABLES_REGION_KEY)
-        if len(set(all_region_keys)) != 1:
-            raise RuntimeError("Tables have incompatible region keys")
-        merged_region_key = all_region_keys[0]
-
-        all_regions = []
-        for table in tables:
-            region = table.uns[TableModel.ATTRS_KEY][TableModel.REGION_KEY]
-            if isinstance(region, str):
-                all_regions.append(region)
-            else:
-                all_regions.extend(region)
-        all_regions = list(set(all_regions))
-        merged_region = all_regions
-
-    attr = {"region": merged_region, "region_key": merged_region_key, "instance_key": merged_instance_key}
-    merged_table = anndata.concat(tables, join="outer", uns_merge="same")
-
-    # remove the MERGED_TABLES_REGION_KEY column if it has been added (the code above either adds that column
-    # to all the tables, either it doesn't add it at all)
-    for table in tables:
-        if MERGED_TABLES_REGION_KEY in table.obs:
-            del table.obs[MERGED_TABLES_REGION_KEY]
-
-    merged_table.uns[TableModel.ATTRS_KEY] = attr
-    merged_table.obs[merged_region_key] = merged_table.obs[merged_region_key].astype("category")
-    TableModel().validate(merged_table)
-    return merged_table
-
+        if region_key is None:
+            raise ValueError("`region_key` must be specified if tables have different region keys")
 
-def concatenate(sdatas: list[SpatialData], omit_table: bool = False) -> SpatialData:
-    """Concatenate a list of spatial data objects.
+    if len(set(instance_keys)) == 1:
+        instance_key = list(instance_keys)[0]
+    else:
+        if instance_key is None:
+            raise ValueError("`instance_key` must be specified if tables have different instance keys")
+
+    tables_l = []
+    regions_l = []
+    instance_keys_l = []
+    for k, i, regs, table in zip(region_keys, instance_keys, regions, tables):
+        if isinstance(regs, list):
+            for r in regs:
+                table_ = table[table.obs[k] == r]
+                tables_l.append(table_)
+                regions_l.append(r)
+                instance_keys_l.append(table_.obs[i].values.flatten())
+        else:
+            tables_l.append(table)
+            regions_l.append(regs)
+            instance_keys_l.append(table.obs[i].values.flatten())
+
+    merged_table = anndata.concat(tables_l, label=region_key, keys=regions_l, **kwargs)
+    merged_table.obs[instance_key] = np.concatenate(instance_keys_l, axis=0)
```

Comment:
Maybe something more like this?

```suggestion
    from copy import copy  # Should probably go up at the top

    tables_l = []
    for table_region_key, table_instance_key, table in zip(region_keys, instance_keys, tables):
        rename_dict = {}
        if table_region_key != region_key:
            rename_dict[table_region_key] = region_key
        if table_instance_key != instance_key:
            rename_dict[table_instance_key] = instance_key
        if len(rename_dict) > 0:
            table = copy(table)  # Shallow copy
            table.obs = table.obs.rename(columns=rename_dict, copy=False)
        tables_l.append(table)

    merged_table = ad.concat(tables_l, **kwargs)
```

* I found the current version took a little effort to figure out. If you want to keep it, it could use a comment explaining that it's making the column names consistent
* This should be more performant, as it doesn't require the subsetting, and the shallow copy should be close to free

The result is slightly different. The order of the observations in tables is preserved, and the result couldn't include the original region or instance keys. If you wanted the possibility of preserving all the original columns in an `outer` join, you could add copies of the instance and region key columns instead of renaming them.